Remove unnecessary dependencies and refactor JSON handling

ruizlenato opened the pull request:
This PR removes unnecessary dependencies and improves JSON handling by replacing third-party libraries with standard library solutions.

**Changes:**
- **Testing Improvements:**
1. Remove stretchr/testify dependency.
2. Replace assert-based tests with direct error checking in player_test.go.
3. Simplify assertions using standard testing package.

- **Dependency Cleanup & Refactoring:**
1. Remove github.com/corpix/uarand and implement a custom RandomUserAgent().
2. Replace valyala/fastjson with encoding/json in getVisitorData().
3. Remove github.com/bitly/go-simplejson and introduce structured types for playlist parsing.
4. Cleanup go.mod/go.sum to remove unused dependencies.

**These changes reduce external dependencies, improve maintainability, and simplify error handling.**

ruizlenato committed:
refactor: replace simplejson with structured types for playlist parsing

- Remove dependency on `github.com/bitly/go-simplejson`
- Introduce strongly-typed structs in `playlist_types.go` for JSON unmarshaling
- Rewrite `parsePlaylistInfo` and related functions to use new types
- Remove unused JSON helper functions (`getFirstKeyJSON`, `isValidJSON`, etc.)
- Simplify continuation token handling and playlist entry extraction

ruizlenato committed:
refactor: replace fastjson with standard library json

- Remove valyala/fastjson dependency and use encoding/json instead
- Update getVisitorData() to use native JSON parsing with proper error handling
- Clean up go.mod/go.sum by removing unused dependencies
- Add proper error checking for io.ReadAll in getVisitorData()

ruizlenato committed:
refactor: remove uarand dependency and implement custom user agent handling

- Remove github.com/corpix/uarand dependency from go.mod/go.sum
- Replace uarand.GetRandom() with custom RandomUserAgent() implementation
- Clean up unused dependencies in project

ruizlenato committed:
chore: remove test dependencies and simplify test assertions

- Remove stretchr/testify dependency from go.mod/go.sum
- Replace assert-based tests with direct error checking in player_test.go
- Clean up remaining test-related dependencies in go.mod
- Simplify test assertions using standard library testing methods

ruizlenato committed:
refactor: simplify code and modernize type declarations

- Remove unused client parameter from parsePlaylistInfo call
- Replace interface{} with any for better readability
- Simplify nsig_function check by removing redundant nil check
- Clean up function signatures and parameter passing

steino commented:
GetPlaylist seems to fail with these changes:

> GetPlaylist (WEB Client)
> === RUN   TestPlaylist
>     playlist_test.go:18: JSON parsing error: runtime error: index out of range [0] with length 0
>         goroutine 22 [running]:
>         runtime/debug.Stack()
>                 /usr/local/go/src/runtime/debug/stack.go:26 +0x5e
>         github.com/steino/youtubedl.(*Playlist).parsePlaylistInfo.func1()
>                 /home/steino/dev/go/youtubedl/playlist.go:62 +0x25
>         panic({0xb16260?, 0xc00278cc00?})
>                 /usr/local/go/src/runtime/panic.go:785 +0x132
>         github.com/steino/youtubedl.(*Playlist).parsePlaylistInfo(0xc000324000, {0xc59de8, 0xc0028eba40}, {0xc002afa000, 0x559db, 0x56000})
>                 /home/steino/dev/go/youtubedl/playlist.go:80 +0xb8f
>         github.com/steino/youtubedl.(*Client).GetPlaylistContext(0xc001c50140, {0xc59d40, 0x1235e60}, {0xb7180a, 0x22}, {0x0, 0x0, 0x53ffe0?})
>                 /home/steino/dev/go/youtubedl/client.go:283 +0x771
>         github.com/steino/youtubedl.(*Client).GetPlaylist(...)
>                 /home/steino/dev/go/youtubedl/client.go:231
>         github.com/steino/youtubedl_test.TestPlaylist(0xc0001351e0)
>                 /home/steino/dev/go/youtubedl/playlist_test.go:16 +0x9d
>         testing.tRunner(0xc0001351e0, 0xba5688)
>                 /usr/local/go/src/testing/testing.go:1690 +0xf4
>         created by testing.(*T).Run in goroutine 1
>                 /usr/local/go/src/testing/testing.go:1743 +0x390
> 
> --- FAIL: TestPlaylist (1.14s)

> GetPlaylist (ANDROID/IOS Client)
> === RUN   TestPlaylist
>     playlist_test.go:18: playlistMetadataRenderer not found in json body
> --- FAIL: TestPlaylist (1.18s)

ruizlenato commented:
Where's this playlist_test?

steino commented:
I haven't made a full test-suite for playlist yet, this is just a simple:

```
c, err := youtubedl.New()
if err != nil {
        t.Error(err)
}

pl, err := c.GetPlaylist(id)
if err != nil {
        t.Error(err)
}
```

ruizlenato commented:
Could you send me the ID you tested? I tested it here and it works normally.

ruizlenato committed:
Fix: Handle nullable video durations in playlist entries

- Changed `Duration` in `PlaylistEntry` to a pointer type (`*time.Duration`).
- Updated `PlaylistVideoRenderer.LengthSeconds` to be `*string` to support missing values.
- Modified `PlaylistEntry()` to safely handle `nil` durations instead of panicking.

ruizlenato commented:
![image](https://github.com/user-attachments/assets/f7694d28-6d4b-47d6-8a56-5fadd0d5a1cd)

Could you send me the ID you tested? I tested it here and it works normally.

steino commented:
You can try this ID: PLwYhoqpb8ldbuLS0k_EoOp06PnG0G271K

ruizlenato commented:
> You can try this ID: PLwYhoqpb8ldbuLS0k_EoOp06PnG0G271K

![image](https://github.com/user-attachments/assets/5cb1a951-daa9-49cb-991b-934ba3f6fe46)

**my go.mod:**"
![image](https://github.com/user-attachments/assets/ada1c363-0594-445d-a803-41e4c893b285)

ruizlenato commented:
any updates?

steino commented:
I tested again on a different machine, fresh go install and fresh git checkout and I still get the same error:

> $ go test -run ^TestPlaylist$ -test.v
> go: downloading github.com/patrickmn/go-cache v2.1.0+incompatible
> go: downloading github.com/mengzhuo/cookiestxt v1.0.4
> go: downloading github.com/dop251/goja v0.0.0-20250125213203-5ef83b82af17
> go: downloading golang.org/x/exp v0.0.0-20250128182459-e0ece0dbea4c
> go: downloading github.com/go-sourcemap/sourcemap v2.1.3+incompatible
> go: downloading github.com/dlclark/regexp2 v1.11.4
> go: downloading golang.org/x/text v0.3.8
> go: downloading github.com/google/pprof v0.0.0-20230207041349-798e818bf904
> === RUN   TestPlaylist
>     playlist_test.go:18: JSON parsing error: runtime error: index out of range [0] with length 0
>         goroutine 8 [running]:
>         runtime/debug.Stack()
>                 /usr/local/go/src/runtime/debug/stack.go:26 +0x5e
>         github.com/steino/youtubedl.(*Playlist).parsePlaylistInfo.func1()
>                 /home/steino/dev/youtubedl/playlist.go:62 +0x25
>         panic({0xb32720?, 0xc002cf04c8?})
>                 /usr/local/go/src/runtime/panic.go:792 +0x132
>         github.com/steino/youtubedl.(*Playlist).parsePlaylistInfo(0xc0003bc420, {0xc71140, 0xc0030f4fc0}, {0xc0007fc000, 0x43c32, 0x44000})
>                 /home/steino/dev/youtubedl/playlist.go:80 +0xb8f
>         github.com/steino/youtubedl.(*Client).GetPlaylistContext(0xc0030fa060, {0xc71098, 0x129ec40}, {0xb91557, 0x22}, {0xc00134bf30, 0x1, 0xd81492?})
>                 /home/steino/dev/youtubedl/client.go:283 +0x771
>         github.com/steino/youtubedl.(*Client).GetPlaylist(...)
>                 /home/steino/dev/youtubedl/client.go:231
>         github.com/steino/youtubedl_test.TestPlaylist(0xc0000be8c0)
>                 /home/steino/dev/youtubedl/playlist_test.go:16 +0xdf
>         testing.tRunner(0xc0000be8c0, 0xbb0ee0)
>                 /usr/local/go/src/testing/testing.go:1792 +0xf4
>         created by testing.(*T).Run in goroutine 1
>                 /usr/local/go/src/testing/testing.go:1851 +0x413
> 
> --- FAIL: TestPlaylist (0.88s)
> FAIL
> exit status 1
> FAIL    github.com/steino/youtubedl     0.882s

ruizlenato commented:
> GetPlaylist seems to fail with these changes:
> 
> > GetPlaylist (WEB Client)
> > === RUN   TestPlaylist
> > playlist_test.go:18: JSON parsing error: runtime error: index out of range [0] with length 0
> > goroutine 22 [running]:
> > runtime/debug.Stack()
> > /usr/local/go/src/runtime/debug/stack.go:26 +0x5e
> > github.com/steino/youtubedl.(*Playlist).parsePlaylistInfo.func1()
> > /home/steino/dev/go/youtubedl/playlist.go:62 +0x25
> > panic({0xb16260?, 0xc00278cc00?})
> > /usr/local/go/src/runtime/panic.go:785 +0x132
> > github.com/steino/youtubedl.(*Playlist).parsePlaylistInfo(0xc000324000, {0xc59de8, 0xc0028eba40}, {0xc002afa000, 0x559db, 0x56000})
> > /home/steino/dev/go/youtubedl/playlist.go:80 +0xb8f
> > github.com/steino/youtubedl.(*Client).GetPlaylistContext(0xc001c50140, {0xc59d40, 0x1235e60}, {0xb7180a, 0x22}, {0x0, 0x0, 0x53ffe0?})
> > /home/steino/dev/go/youtubedl/client.go:283 +0x771
> > github.com/steino/youtubedl.(*Client).GetPlaylist(...)
> > /home/steino/dev/go/youtubedl/client.go:231
> > github.com/steino/youtubedl_test.TestPlaylist(0xc0001351e0)
> > /home/steino/dev/go/youtubedl/playlist_test.go:16 +0x9d
> > testing.tRunner(0xc0001351e0, 0xba5688)
> > /usr/local/go/src/testing/testing.go:1690 +0xf4
> > created by testing.(*T).Run in goroutine 1
> > /usr/local/go/src/testing/testing.go:1743 +0x390
> > --- FAIL: TestPlaylist (1.14s)
> 
> > GetPlaylist (ANDROID/IOS Client)
> > === RUN   TestPlaylist
> > playlist_test.go:18: playlistMetadataRenderer not found in json body
> > --- FAIL: TestPlaylist (1.18s)

I ran a test on my server where YouTube blocks the IP because it's a VPS and I got exactly the same error.

```bash
teste ➜ go run main.go
go: downloading go1.24.1 (linux/arm64)
go: downloading github.com/ruizlenato/youtubedl v0.0.0-20250331230612-2ca51f2e7e73
go: downloading github.com/dop251/goja v0.0.0-20250309171923-bcd7cc6bf64c
go: downloading github.com/mengzhuo/cookiestxt v1.0.4
go: downloading github.com/patrickmn/go-cache v2.1.0+incompatible
go: downloading golang.org/x/exp v0.0.0-20250305212735-054e65f0b394
go: downloading github.com/go-sourcemap/sourcemap v2.1.4+incompatible
go: downloading golang.org/x/text v0.23.0
go: downloading github.com/dlclark/regexp2 v1.11.5
go: downloading github.com/google/pprof v0.0.0-20250317173921-a4b03ec1a45e
2025/04/05 11:41:24 JSON parsing error: runtime error: index out of range [0] with length 0
goroutine 1 [running]:
runtime/debug.Stack()
        /home/renato/.local/share/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-arm64/src/runtime/debug/stack.go:26 +0x64
github.com/steino/youtubedl.(*Playlist).parsePlaylistInfo.func1()
        /home/renato/.local/share/go/pkg/mod/github.com/ruizlenato/youtubedl@v0.0.0-20250331230612-2ca51f2e7e73/playlist.go:62 +0x24
panic({0x61d680?, 0x400241e4c8?})
        /home/renato/.local/share/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-arm64/src/runtime/panic.go:792 +0x124
github.com/steino/youtubedl.(*Playlist).parsePlaylistInfo(0x4000580180, {0x7429b0, 0x400275edb0}, {0x400158c000, 0x45472, 0x56000})
        /home/renato/.local/share/go/pkg/mod/github.com/ruizlenato/youtubedl@v0.0.0-20250331230612-2ca51f2e7e73/playlist.go:80 +0x958
github.com/steino/youtubedl.(*Client).GetPlaylistContext(0x400275a500, {0x742898, 0xd13ca0}, {0x674637, 0x22}, {0x400041df08, 0x1, 0x50ff0?})
        /home/renato/.local/share/go/pkg/mod/github.com/ruizlenato/youtubedl@v0.0.0-20250331230612-2ca51f2e7e73/client.go:283 +0x5dc
github.com/steino/youtubedl.(*Client).GetPlaylist(...)
        /home/renato/.local/share/go/pkg/mod/github.com/ruizlenato/youtubedl@v0.0.0-20250331230612-2ca51f2e7e73/client.go:231
main.main()
        /home/renato/teste/main.go:16 +0xb4
exit status 1
```

Now that I know how to reproduce the error, I'll try to fix it.

ruizlenato committed:
fix: add bounds checking for playlist author extraction

ruizlenato commented:
Fixed

> GetPlaylist seems to fail with these changes:
> 
> > GetPlaylist (WEB Client)
> > === RUN   TestPlaylist
> > playlist_test.go:18: JSON parsing error: runtime error: index out of range [0] with length 0
> > goroutine 22 [running]:
> > runtime/debug.Stack()
> > /usr/local/go/src/runtime/debug/stack.go:26 +0x5e
> > github.com/steino/youtubedl.(*Playlist).parsePlaylistInfo.func1()
> > /home/steino/dev/go/youtubedl/playlist.go:62 +0x25
> > panic({0xb16260?, 0xc00278cc00?})
> > /usr/local/go/src/runtime/panic.go:785 +0x132
> > github.com/steino/youtubedl.(*Playlist).parsePlaylistInfo(0xc000324000, {0xc59de8, 0xc0028eba40}, {0xc002afa000, 0x559db, 0x56000})
> > /home/steino/dev/go/youtubedl/playlist.go:80 +0xb8f
> > github.com/steino/youtubedl.(*Client).GetPlaylistContext(0xc001c50140, {0xc59d40, 0x1235e60}, {0xb7180a, 0x22}, {0x0, 0x0, 0x53ffe0?})
> > /home/steino/dev/go/youtubedl/client.go:283 +0x771
> > github.com/steino/youtubedl.(*Client).GetPlaylist(...)
> > /home/steino/dev/go/youtubedl/client.go:231
> > github.com/steino/youtubedl_test.TestPlaylist(0xc0001351e0)
> > /home/steino/dev/go/youtubedl/playlist_test.go:16 +0x9d
> > testing.tRunner(0xc0001351e0, 0xba5688)
> > /usr/local/go/src/testing/testing.go:1690 +0xf4
> > created by testing.(*T).Run in goroutine 1
> > /usr/local/go/src/testing/testing.go:1743 +0x390
> > --- FAIL: TestPlaylist (1.14s)
> 
> > GetPlaylist (ANDROID/IOS Client)
> > === RUN   TestPlaylist
> > playlist_test.go:18: playlistMetadataRenderer not found in json body
> > --- FAIL: TestPlaylist (1.18s)

I think it's been fixed
